python/python/bystro/covariance/_base_covariance.py

```diff
@@ -0,0 +1,549 @@
+
+    beta_bar = la.solve(covariance_22, covariance_21)
+    Second_part = np.dot(covariance_21.T, beta_bar)
+
+    covariance_11 - Second_part
+
+    preds = np.dot(Xobs[:, idxs == 1], beta_bar)
+    return preds
+
+
+#########################################
+#########################################
+##                                     ##
+##  Gaussian Likelihood-based methods  ##
+##                                     ##
+#########################################
+#########################################
+
+
+def _conditional_score(covariance, X, idxs, weights=None):
+    """
+    Returns the predictive log-likelihood of a subset of data.
+
+    mean(log p(X[idx==1]|X[idx==0],covariance))
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(idxs))
+        The data
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    weights : np.array-like,(N,),default=None
+        The optional weights on the samples
+
+    Returns
+    -------
+    avg_score : float
+        Average log likelihood
+    """
+    if weights is None:
+        weights = np.ones(X.shape[0])
+    avg_score = np.mean(
+        weights * _conditional_score_samples(covariance, X, idxs)
+    )
+    return avg_score
+
+
+def _conditional_score_samples(covariance, X, idxs):
+    """
+    Return the conditional log likelihood of each sample, that is
+
+    log p(X[idx==1]|X[idx==0],covariance) = N(Sigma_10Sigma_00^{-1}x_0,
+                                    Sigma_11-Sigma_10Sigma_00^{-1}Sigma_01)
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,p)
+        The data
+
+    idxs: np.array-like,(p,)
+        The observation locations
+
+    Returns
+    -------
+    scores : float
+        Log likelihood for each sample
+    """
+    covariance_sub = covariance[idxs == 1]
+    covariance_22 = covariance_sub[:, idxs == 1]
+    covariance_21 = covariance_sub[:, idxs == 0]
+
+    covariance_nonsub = covariance[idxs == 0]
+    covariance_11 = covariance_nonsub[:, idxs == 0]
+
+    beta_bar = la.solve(covariance_22, covariance_21)
+    Second_part = np.dot(covariance_21.T, beta_bar)
+
+    covariance_bar = covariance_11 - Second_part
+
+    mu_ = np.dot(X[:, idxs == 1], beta_bar)
+    scores = _score_samples(covariance_bar, X[:, idxs == 0] - mu_)
+
+    return scores
+
+
+def _conditional_parameters(covariance, idxs):
+    """
+    Computes the distribution parameters p(X_miss|X_obs)
+
+    Parameters
+    ----------
+    covariance : np.array-like,(p,p)
+        The covariance matrix
+
+    idxs: np.array-like,(p,)
+        The observed covariates
+
+    Returns
+    -------
+    beta_bar : array-like
+        The predictive covariates
+
+    covariance_bar : array-like
+        Conditional covariance
+    """
+
+    covariance_sub = covariance[idxs == 1]
```

Comment:
looks like we should really factor this block out

Reply:
What do you mean?

Reply:
Can't for reason mentioned above. Unfortunately inverses don't work out the way you think that would allow this.

Reply:
Could you say more about that (maybe in the thread above for consolidation): https://github.com/bystrogenomics/bystro/pull/215/files#r1315087573?  I'm worried that we're talking past each other here (e.g. I'm not claiming that inverses work in any particular way, I'm just noticing a pattern of code duplication).

Reply:
Sigma_11^{-1} is not the same as Phi_11 where Phi=Sigma^{-1}. Can test out with a matrix generated by ortho_group where Sigma_11^{-1} != I = Phi_11